before_after_filesystem_snapshot.py

```diff
@@ -0,0 +1,64 @@
+def snapshot(before_dict, after_dict):
+  '''before_after_snapshot is a simple function that returns which files were
+    unchanged, modified, added or removed from an input dictionary (before_dict)
+    and an output dictionary (after_dict). Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = []
+  modified_files = []
+  added_files = []
+  removed_files = []
+  for key in before_dict:
+    if key in after_dict:
+      if before_dict[key] == after_dict[key]:
+        # Matching the hashes to check if file was unchanged
+        unchanged_files.append(key)
+      else:
+        modified_files.append(key)
+    else:
+      removed_files.append(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+  sorted(removed_files))
+
+def generate_artifact_rules(snapshot):
+  '''
+  Generate Artifact Rules given which files have been added, which have been removed,
+  which have been modified, and which have remained unchanged. 
+  '''
+  artifact_rules = {
+    'expected_materials': [],
+    'expected_products': []
+  }
+
+  # missing rules for MATCH since we don't have the information of the material from 
+  # the previous step
+  for file in snapshot[0]:
+    # unchanged files
+    artifact_rules['expected_materials'].append(["ALLOW", file])
+  for file in snapshot[1]:
+    # modified files
+    artifact_rules['expected_materials'].append(["ALLOW", file])
+  artifact_rules['expected_materials'].append(["DISALLOW", "*"])
+
+  for file in snapshot[0]:
+    # unchanged files
+    artifact_rules['expected_products'].append(["ALLOW", file])
+  for file in snapshot[1]:
+    # modified files
+    artifact_rules['expected_products'].append(["MODIFY", file])
+  for file in snapshot[2]:
+    # added files
+    artifact_rules['expected_products'].append(["CREATE", file])
+  for file in snapshot[3]:
+    # removed files
+    artifact_rules['expected_products'].append(["DELETE", file])
```

Comment:
Note that the delete rule does not have an effect if it is a product rule. This is because product rules can only consume artifacts from the product queue, but a deleted artifact won't show up in the product queue, it only shows up in the materials queue, so it must be consumed there.

For the create rule it's the other way around.

(cross-referencing https://github.com/in-toto/docs/issues/4)

Reply:
Let me know if this isn't clear.

Reply:
Thanks a lot for the review! @lukpueh 

I think a summary of what you mean is that `DELETE` rule should be included in materials rules and `CREATE` rule should be included in products rules?

I understand your point about the `DELETE` rule now. I am also thinking that is `DISALLOW *` in products rules do a similar job as `DELETE` rule? Since `DISALLOW *` indicates that all other artifacts are not allowed as products, does it also indicates that deleted artifacts should not show up in the products queue?

Reply:
> I think a summary of what you mean is that DELETE rule should be included in materials rules and CREATE rule should be included in products rules?

Exactly! :)

> I am also thinking that is DISALLOW * in products rules do a similar job as DELETE rule?

They are actually quite different. DELETE  is a rule that consumes artifacts, i.e. removes it from the queue when matched by the pattern. Since deleted files can only be in the materials queue, which is only being looked at when processing material rules, they can also only be consumed there.
The DISALLOW rule, on the other hand, does not consume artifacts. Instead it fails the verification if its pattern matches any artifacts in the corresponding queue (material queue if the DISALLOW rule is in expected_materials, and product queue if the DISALLOW rule is in the expected_products).